[🐛 Bug]: Randomly getting java.lang.ClassCastException when trying to interact with a particular element

### Description

- after loading the page i am trying to click on a login link
- ⁠i expect it to click
- ⁠getting the following exception
```java
REST API Reason:java.lang.ClassCastException: 
  class java.util.LinkedHashMap cannot be cast to class org.openqa.selenium.WebElement
    (java.util.LinkedHashMap is in module java.base of loader 'bootstrap';
     org.openqa.selenium.WebElement is in unnamed module of loader...
```

- stacktrace : 
```java
java.lang.ClassCastException: class java.util.LinkedHashMap cannot be cast to class org.openqa.selenium.WebElement (java.util.LinkedHashMap is in module java.base of loader 'bootstrap'; org.openqa.selenium.WebElement is in unnamed module of loader 'app')
	at org.openqa.selenium.remote.ElementLocation$ElementFinder$2.findElement(ElementLocation.java:166)
	at org.openqa.selenium.remote.ElementLocation.findElement(ElementLocation.java:59)
	at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:367)
	at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:361)
	at org.openqa.selenium.support.ui.ExpectedConditions$7.apply(ExpectedConditions.java:194)
	at org.openqa.selenium.support.ui.ExpectedConditions$7.apply(ExpectedConditions.java:190)
	at org.openqa.selenium.support.ui.ExpectedConditions$22.apply(ExpectedConditions.java:627)
	at org.openqa.selenium.support.ui.ExpectedConditions$22.apply(ExpectedConditions.java:624)
	at org.openqa.selenium.support.ui.FluentWait.until(FluentWait.java:203)
```


### Reproducible Code

```java
public void tapLoginButton(WebDriver driver) {
    By loginLocator = By.xpath("//p[contains(text(),'Login')]");
    safeClick(driver, loginLocator, 10);
}

public void safeClick(WebDriver driver, By locator, int timeoutSeconds) {
    WebElement element = waitForElementClickable(driver, locator, timeoutSeconds);
    element.click();
}

public WebElement waitForElementClickable(WebDriver driver, By locator, int timeoutSeconds) {
    WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(timeoutSeconds));
    return wait.until(ExpectedConditions.elementToBeClickable(locator));
}
```

**Other Information** 

1. Locally does not seem to be happening
2. Java version 11
3. Observing this randomly on this particular point of our script on trying to interact with this `<p class="D_le M_lK D_lf M_lL D_lj M_lP D_ll M_lR D_lp M_lV D_ls M_lY D_la" style="color:unset">Login</p>`
4. We tried this on browserstack chrome version 138/140 and OS 10 and 11
5. sorry, cant share the URL
6. We tried without waiting for the element to be clickable by directly clicking on the element, still got the same error

@joerg1985 @diemol I managed to reproduce the issue with latest Selenium 4.39.0 and [beforementioned code](https://github.com/SeleniumHQ/selenium/issues/16389#issuecomment-3372822082).

#### The problem
The problem is that Selenium expects remote response to contain both "error" and "message" elements, while BrowserStack contains only "message":

```json
{
  value: {
    message: "Session not started or terminated"
  },
  sessionId: "",
  status: 13
}
```

Seems that the problem happens only with few specific responses a'la "session not started".
In most cases, BrowserStack contains both "error" and "message":

```json
{
  value: {
    error: "no such element",
    message: "no such element: Unable to locate element: {\"method\":\"xpath\", \"selector\": \"//p[contains(text(),'Login')]\"}"
  }
}
```

#### The full error stacktrace:

```java
org.openqa.selenium.remote.UnreachableBrowserException: Error communicating with the remote browser. It may have died.
Build info: version: '4.39.0', revision: '126f156aee'
System info: os.name: 'Mac OS X', os.arch: 'aarch64', os.version: '26.2', java.version: '17.0.17'
Driver info: org.openqa.selenium.remote.RemoteWebDriver$ByteBuddy$2bw6qaSS$ByteBuddy$0LakB7vv
Command: [2eaeaabfbc53c2cecf75a3cb63a65c5db9a0097e, findElement [using, value]]

	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:592)
	at org.openqa.selenium.remote.ElementLocation$ElementFinder$2.findElement(ElementLocation.java:166)
	at org.openqa.selenium.remote.ElementLocation.findElement(ElementLocation.java:60)
	at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:384)
	at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:378)
	at org.selenide.bs.LinkedHashMapBugReproducerTest.issue16389(LinkedHashMapBugReproducerTest.java:23)

Caused by: java.lang.IllegalArgumentException: Unable to find mapping for {value={message=Session not started or terminated}, sessionId=, status=13}
	at org.openqa.selenium.remote.ErrorCodec.decode(ErrorCodec.java:152)
	at org.openqa.selenium.remote.codec.w3c.W3CHttpResponseCodec.decode(W3CHttpResponseCodec.java:142)
	at org.openqa.selenium.remote.codec.w3c.W3CHttpResponseCodec.decode(W3CHttpResponseCodec.java:49)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:215)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:561)
	... 5 more
```

#### My verdict
It's a BrowerStack's issue, but we can create some workaround on Selenium side.

Probably this workaround wouldn't give any big benefit because currently Selenium ends up with `UnreachableBrowserException` which is already very close to the reality. :)

--- a/java/src/org/openqa/selenium/remote/ErrorCodec.java
+++ b/java/src/org/openqa/selenium/remote/ErrorCodec.java
@@ -17,10 +17,14 @@
 
 package org.openqa.selenium.remote;
 
+import static java.util.Objects.requireNonNullElse;
+import static java.util.Objects.requireNonNullElseGet;
+
 import java.io.PrintWriter;
 import java.io.StringWriter;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.function.Function;
 import org.openqa.selenium.DetachedShadowRootException;
 import org.openqa.selenium.ElementClickInterceptedException;
@@ -88,7 +92,7 @@ public class ErrorCodec {
           new W3CError<>("unsupported operation", UnsupportedCommandException::new, 500),
           new W3CError<>("unknown command", UnsupportedCommandException::new, 404),
           new W3CError<>("unknown method", UnsupportedCommandException::new, 405),
-          new W3CError<>("unknown error", WebDriverException::new, 500));
+          DEFAULT_ERROR);
 
   private ErrorCodec() {
     // This will switch to being an interface at some point. Use `createDefault`
@@ -144,24 +148,28 @@ public class ErrorCodec {
 
   public WebDriverException decode(Map<String, Object> response) {
     if (!(response.get("value") instanceof Map)) {
-      throw new IllegalArgumentException("Unable to find mapping for " + response);
+      throw new InvalidResponseException("missing \"value\" field", response);
     }
 
     Map<?, ?> value = (Map<?, ?>) response.get("value");
-    if (!(value.get("error") instanceof String)) {
-      throw new IllegalArgumentException("Unable to find mapping for " + response);
+
+    Object error = requireNonNullElse(value.get("error"), "");
+    Object message = requireNonNullElseGet(value.get("message"), response::toString);
+
+    if (!(error instanceof String)) {
+      throw new InvalidResponseException("\"error\" field must be a string", response);
+    }
+    if (!(message instanceof String)) {
+      throw new InvalidResponseException("\"message\" field must be a string", response);
     }
 
-    String error = (String) value.get("error");
-    String message = value.get("message") instanceof String ? (String) value.get("message") : null;
-
-    W3CError<?> w3CError =
-        ERRORS.stream()
-            .filter(err -> error.equals(err.w3cErrorString))
-            .findFirst()
-            .orElse(DEFAULT_ERROR);
-
-    return w3CError.exceptionConstructor.apply(message);
+    Optional<W3CError<? extends WebDriverException>> w3CError =
+        ERRORS.stream().filter(err -> error.equals(err.w3cErrorString)).findFirst();
+    if (w3CError.isPresent()) {
+      return w3CError.get().exceptionConstructor.apply((String) message);
+    }
+    String extendedMessage = String.format("%s (error code: \"%s\")", message, error);
+    return DEFAULT_ERROR.exceptionConstructor.apply(extendedMessage);
   }
 
   private W3CError<? extends WebDriverException> fromThrowable(Throwable throwable) {

--- a/java/src/org/openqa/selenium/remote/InvalidResponseException.java
+++ b/java/src/org/openqa/selenium/remote/InvalidResponseException.java
@@ -1,0 +1,30 @@
+// Licensed to the Software Freedom Conservancy (SFC) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The SFC licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.openqa.selenium.remote;
+
+import java.util.Map;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public class InvalidResponseException extends IllegalArgumentException {
+  private static final String w3cErrorFormat = "https://www.w3.org/TR/webdriver2/#errors";
+
+  public InvalidResponseException(String message, Map<String, Object> response) {
+    super(String.format("%s: %s%nSee %s", message, response, w3cErrorFormat));
+  }
+}

--- a/java/test/org/openqa/selenium/remote/ErrorCodecTest.java
+++ b/java/test/org/openqa/selenium/remote/ErrorCodecTest.java
@@ -17,12 +17,14 @@
 
 package org.openqa.selenium.remote;
 
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
 import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.text.ParseException;
 import java.util.Map;
 import java.util.UUID;
-import org.assertj.core.api.Assertions;
 import org.junit.jupiter.api.Tag;
 import org.junit.jupiter.api.Test;
 import org.openqa.selenium.DetachedShadowRootException;
@@ -119,11 +121,53 @@ public class ErrorCodecTest {
     encodeAndCheck(new ClassNotFoundException(msg), "unknown error", msg);
   }
 
+  @Test
+  void decode_w3cCompliantResponse() {
+    assertThat(
+            errorCodec.decode(
+                Map.of(
+                    "value",
+                    Map.of(
+                        "error", "script timeout",
+                        "message", "Script timed out"))))
+        .isInstanceOf(ScriptTimeoutException.class)
+        .hasMessageStartingWith("Script timed out");
+  }
+
+  @Test
+  void decode_unknownError() {
+    assertThat(
+            errorCodec.decode(
+                Map.of(
+                    "value",
+                    Map.of(
+                        "error", "some.magic.error.code",
+                        "message", "Something unbelievable happened"))))
+        .isInstanceOf(WebDriverException.class)
+        .hasMessageStartingWith("Something unbelievable happened")
+        .hasMessageContaining("some.magic.error.code");
+  }
+
+  @Test
+  void decode_noValue() {
+    assertThatThrownBy(() -> errorCodec.decode(Map.of("unexpected", "format")))
+        .isInstanceOf(InvalidResponseException.class)
+        .hasMessageStartingWith("missing \"value\" field: {unexpected=format}")
+        .hasMessageContaining("See https://www.w3.org/TR/webdriver2/#errors");
+  }
+
+  @Test
+  void decode_noError() {
+    assertThat(errorCodec.decode(Map.of("value", Map.of("message", "Session was not created"))))
+        .isInstanceOf(WebDriverException.class)
+        .hasMessageStartingWith("Session was not created");
+  }
+
   void encodeAndCheck(Throwable toEncode, String expectedType, String expectedMessage) {
     Map<String, Object> encoded = errorCodec.encode(toEncode);
     Map<String, Object> value = (Map<String, Object>) encoded.get("value");
 
-    Assertions.assertThat(value.get("error")).isEqualTo(expectedType);
-    Assertions.assertThat(value.get("message")).isEqualTo(toEncode.getMessage());
+    assertThat(value.get("error")).isEqualTo(expectedType);
+    assertThat(value.get("message")).isEqualTo(toEncode.getMessage());
   }
 }

